Add link for docs, show changelog

Martchus opened the pull request:
See https://progress.opensuse.org/issues/16062

![spectacle t25241](https://cloud.githubusercontent.com/assets/10248953/22063567/c5c69818-dd7f-11e6-97c8-b9f7d7fb90c1.png)

Still to improve:
* formatting
* more generic approach for getting change logs or at least handle case where logs can not be queried via `rpm`
* maybe some caching
* also show os-autoinst log for worker

okurz commented:
I think that can not be a generic approach so I recommend to put that into branding.

> also show os-autoinst log for worker

won't work because you can not ensure workers all have the same os-autoinst version. Better don't invest time on that. That is an administration issue, e.g. to be solved by salt.

Martchus commented:
Of course it would have to show the os-autoinst version for each worker specifically, eg. on worker page. This information could be provided by the worker when registering. But yes, it might not be required.

foursixnine commented:
@Martchus, let's wait until @okurz improves the user story and then we can decide what to do from there. I like the idea to have the worker Changelog on the worker page.

coolo commented on lib/OpenQA/WebAPI/Controller/Main.pm:
```diff
@@ -162,5 +163,12 @@ sub parent_group_overview {
 
+sub changelog {
+    my ($self) = @_;
+    my $changelog = decode('utf8', `rpm -q --changelog openQA`);
```
this needs error handling. openQA does not need to be installed from rpm - i.e. in tests

foursixnine replied:
Actually i submitted a change on the obs project so that we are packaging the openQA.changes file into /public/Changelog.

coolo replied:
you still need error handling if that file is not packaged - i.e. in tests

Martchus replied:
Of course this needs error handling (that's why it is still WIP). But I'll wait with it till the change log is packaged.

coolo commented:
I don't think it should be in branding - but it should be testable and have a fallback if there is no rpm openQA, i.e. mock it in tests.

okurz commented:
On Wednesday, 18 January 2017 08:22:55 CET Stephan Kulow wrote:
> a fallback if there is no rpm openQA, i.e. mock it in tests.

at best show output of `git describe` based on latest git tag release - there 
aren't any - or git hash and fallback to just '(unknown)'

Martchus commented:
@okurz Using `git` commands will not work because when packaged, the files are not in a git repository anymore.

okurz commented:
that's why I quoted coolo's request for fallback. Do like this

if installed from package show package version
else if in a git repo show `git describe`
else show '(unknown)' or nothing at all

foursixnine commented on lib/OpenQA/WebAPI/Controller/Main.pm:
```diff
@@ -162,5 +163,20 @@ sub parent_group_overview {
+    my ($self) = @_;
+
+    my $changelog;
+    if(open(my $changelog_file, '<', '/usr/share/openqa/ChangeLog')) {
```
once [this](https://build.opensuse.org/request/show/451342) gets accepted into OBS, the Changelog will be in public/Changelog

Martchus replied:
When putting it under `public/Changelog` the user will be able to access the raw change log. As you can see in this pull request I intend to embed the change log in the website and apply some formatting. But I guess it is ok to have both variants accessible.

Maybe I should also make the path configurable since it depends on the way openQA is packaged?

okurz commented:
I still say https://github.com/os-autoinst/openQA/pull/1184#issuecomment-273780978

Also, maybe you can try to provide this functionality as a plugin *or* using branding templates. This way the packager can provide a custom implementation by putting a template file within /etc/openqa/...

Martchus commented:
I moved the implementation to a branding. It now shows `git describe` as a fallback. (Still not sure about the benefit of `git describe` fallback because when using the Git version during development I can just call `git describe` myself from command line.)

okurz commented:
On January 31, 2017 10:55:12 AM GMT+01:00, Martchus <notifications@github.com> wrote:
>I moved the implementation to a branding. It now shows `git describe`
>as a fallback. (Still not sure about the benefit of `git describe`
>fallback because when using the Git version during development I can
>just call `git describe` myself from command line.)

Same thing as the admin on rpm installed instances, he can call `rpm -q --changelog` but the user of the web UI can not

Martchus commented:
Maybe the admin can, but for example as openQA developer without admin rights I find it useful to have the log on the web UI.

It fails only because "2 files have unexpected coverage changes not visible in diff" now.

okurz commented:
you did not get my point. I meant: Yes, you can call `git describe` manually *as an instance admin* but the user can not so it is a useful feature to show the current version on the webui regardless of how the instance was provisioned.

okurz commented:
ah, sorry. we misunderstood. I mixed up "changelog" and "display of version" here. Of course `git describe` corresponds to a version, not a changelog.

So the PR as is right now does not really make sense.

I updated the progress tickets accordingly. Better to revisit this and then do it properly, I suggest "version" first for which part of the code here can be used.

@coolo stated that it should not be in branding, I agree. Although, it would not harm if any helper is also usable in a branding template that can be used for overriding. If you would start with putting the version in the footer than it might be easy. Please see https://progress.opensuse.org/issues/16232 first, though

Martchus commented:
> ah, sorry. we misunderstood. I mixed up "changelog" and "display of version" here. Of course git describe corresponds to a version, not a changelog.

Yes, it is mixed. But that is not too bad. My use case of the feature is just to check whether a particular change is already included and the ID of the latest commit is sufficient for this - an entire change log would just make it more convenient.

> stated that it should not be in branding, I agree.

So I will just switch back to the previous approach which I have still on another branch. But that would be the version already present here for some time.

> And I don't like the additional menu.

Why? Any better ideas? The menu seemed pretty obvious since that's where you can find help in almost any application and there's space left in the menu anyways.

okurz commented:
>> And I don't like the additional menu.
> Why? Any better ideas? The menu seemed pretty obvious since that's where you can find help in almost any application and there's space left in the menu anyways.

As discussed in person. We should try to not overwhelm new / unknown / not logged in users with too many options on one screen so better put it into the user menue.

Martchus commented:
Ok, and I guess the same counts for this: https://github.com/os-autoinst/openQA/pull/1220

Martchus commented:
So everybody would be ok with this in the 'Logged in' menu? Just because I do not want to change it again and again.

Martchus commented:
Since only Oliver shared his opinion I implemented it his way. So the changelog is now accessible via the user menu.

@okurz I did not include `git describe` to not mix version and change log. The version can be included later, too of course.

The path is currently hard-coded to `/usr/share/openqa/public/ChangeLog`.

Martchus commented:
@aaannz @coolo Can this be merged now?

coolo commented on lib/OpenQA/WebAPI/Controller/Main.pm:
```diff
@@ -162,5 +163,20 @@ sub parent_group_overview {
+    if (open(my $changelog_file, '<', '/usr/share/openqa/public/ChangeLog')) {
+        read($changelog_file, $changelog, -s $changelog_file);
+        close($changelog_file);
+        $changelog = Mojo::ByteStream->new(Text::Markdown->new->markdown($changelog));
```
why markdown? changelog is just text.

coolo commented on lib/OpenQA/WebAPI/Controller/Main.pm:
```diff
@@ -162,5 +163,20 @@ sub parent_group_overview {
+    my ($self) = @_;
+
+    my $changelog;
+    if (open(my $changelog_file, '<', '/usr/share/openqa/public/ChangeLog')) {
```
this is too specific to our packages, this needs to be configured (by our package). And if no changelog file is defined, the menu entry shouldn't be there.

coolo commented on templates/main/changelog.html.ep:
```diff
@@ -0,0 +1,13 @@
+% layout 'bootstrap';
+% title '';
```
why no title?

coolo commented on templates/main/changelog.html.ep:
```diff
@@ -0,0 +1,13 @@
+%= include 'layouts/info'
+
+% content_for 'ready_function' => begin
+    setupIndexPage();
```
what's in that javascript function required here?

coolo commented on lib/OpenQA/WebAPI/Controller/Main.pm:
```diff
@@ -23,6 +23,7 @@ use OpenQA::Schema::Result::Jobs;
 use OpenQA::BuildResults;
 use OpenQA::Utils;
 use Scalar::Util 'looks_like_number';
+use Encode qw(decode);
```
I can't see this used

Martchus commented:
I guess most things are easy to fix. In my opinion the log looks much better when formatting it as markdown. The alternative would be using the `pre`-tag which doesn't look so nice.

Making the path configurable at build time makes sense of course. I'm just not sure how to do it since we don't have a configure script. The config INI would be an option. Then the packager would have to use `sed` or similar to replace the default. It might also be possible to get the path of the "public" directory from Mojolicious. However, that the changelog file is installed there is also specific to our packaging.

Martchus commented:
I rebased and changed it to use `pre`-tag.

However, how should I proceed regarding the configuration? I mentioned different options in my last comment. I'm asking before doing anything because I don't want to do it twice. There have already been enough amendments in this PR.

okurz commented:
Leaving it an option to the packager to patch the configuration files or override it with a custom version is common I guess so the way to go

Martchus commented:
Ok, it is now configurable via the config file. The maintainer can change the path by patching it. Otherwise, it defaults to `/usr/share/openqa/public/Changelog` which should be ok for most distros anyways.

Martchus commented:
@coolo Do you still request changes? As far as I remember everything you've mentioned is fixed now.

Martchus commented:
I rebased (again), is this ok now?

coolo commented:
I leave the browser tab open until I find the time - sorry
